vault/identity_mfa.go

```diff
@@ -0,0 +1,1676 @@
+	}
+	if c.barrier.Put(ctx, &logical.StorageEntry{
+		Key:   fmt.Sprintf("%s%s/%s", mfaTOTPKeysPrefix, methodID, entityID),
+		Value: val,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Core) fetchTOTPKey(ctx context.Context, methodID, entityID string) (string, error) {
+	entry, err := c.barrier.Get(ctx, fmt.Sprintf("%s%s/%s", mfaTOTPKeysPrefix, methodID, entityID))
+	if err != nil {
+		return "", err
+	}
+	if entry == nil {
+		return "", nil
+	}
+
+	ks := &totpKey{}
+	err = jsonutil.DecodeJSON(entry.Value, ks)
+	if err != nil {
+		return "", err
+	}
+
+	return ks.Key, nil
+}
+
+func (b *SystemBackend) handleMFAGenerateTOTP(ctx context.Context, mConfig *mfa.Config, entityID string) (*logical.Response, error) {
+	var err error
+	var totpConfig *mfa.TOTPConfig
+
+	if b.Core.identityStore == nil {
+		return nil, fmt.Errorf("identity store not set up, cannot service totp mfa requests")
+	}
+
+	switch mConfig.Config.(type) {
+	case *mfa.Config_TOTPConfig:
+		totpConfig = mConfig.Config.(*mfa.Config_TOTPConfig).TOTPConfig
+	default:
+		return logical.ErrorResponse(fmt.Sprintf("unknown MFA config type %q", mConfig.Type)), nil
+	}
+
+	b.Core.identityStore.lock.Lock()
+	defer b.Core.identityStore.lock.Unlock()
+
+	// Read the entity after acquiring the lock
+	entity, err := b.Core.identityStore.MemDBEntityByID(entityID, true)
+	if err != nil {
+		return nil, errwrap.Wrapf(fmt.Sprintf("failed to find entity with ID %q: {{err}}", entityID), err)
+	}
+
+	if entity == nil {
+		return logical.ErrorResponse("invalid entity ID"), nil
+	}
+
+	if entity.MFASecrets == nil {
+		entity.MFASecrets = make(map[string]*mfa.Secret)
+	} else {
+		_, ok := entity.MFASecrets[mConfig.ID]
+		if ok {
+			resp := &logical.Response{}
+			resp.AddWarning(fmt.Sprintf("Entity already has a secret for MFA method %q", mConfig.Name))
+			return resp, nil
+		}
+	}
+
+	keyObject, err := totplib.Generate(totplib.GenerateOpts{
+		Issuer:      totpConfig.Issuer,
+		AccountName: entity.ID,
+		Period:      uint(totpConfig.Period),
+		Digits:      otplib.Digits(totpConfig.Digits),
+		Algorithm:   otplib.Algorithm(totpConfig.Algorithm),
+		SecretSize:  uint(totpConfig.KeySize),
+		Rand:        b.Core.secureRandomReader,
+	})
+	if err != nil {
+		return nil, errwrap.Wrapf(fmt.Sprintf("failed to generate TOTP key for method name %q: {{err}}", mConfig.Name), err)
+	}
+	if keyObject == nil {
+		return nil, fmt.Errorf("failed to generate TOTP key for method name %q", mConfig.Name)
+	}
+
+	totpURL := keyObject.String()
+
+	totpB64Barcode := ""
+	if totpConfig.QRSize != 0 {
+		barcode, err := keyObject.Image(int(totpConfig.QRSize), int(totpConfig.QRSize))
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to generate QR code image: {{err}}", err)
+		}
+
+		var buff bytes.Buffer
+		png.Encode(&buff, barcode)
+		totpB64Barcode = base64.StdEncoding.EncodeToString(buff.Bytes())
+	}
+
+	if err := b.Core.PersistTOTPKey(ctx, mConfig.ID, entity.ID, keyObject.Secret()); err != nil {
```

Comment:
This part should be changed for the login MFA, as described in the RFC:
`We would prefer to avoid additional storage writes as part of logins using MFA.  This is feasible because when perf standbys handle login requests, although the actual verification of credentials is done on the perf standby, the creation of the token is done via an RPC to the active node.  This means that the active node will always be involved in creating tokens, including in this two-phase pre-MFA-verified case.  The active node will hold the unverified token in memory without persisting it to storage.   sys/mfa/validate requests will always be forwarded to the active node.  When they complete successfully, the active node will persist the token and return a standard Auth response.
`